### PR TITLE
Feature/46 create functionality to create a vehicle

### PR DESCRIPTION
**Pull request by fabriciodigennaro**
Created a new endpoint to create a vehicle.
    methods:
    * saveVehicle() => save a new vehicle
    * getVehicleByUserIdandPlate() => return an optional of Vehicle  by plate and UserId
    * vehicleExistsByUserIdAndPlate() => that checks if the vehicle already exists in DB

Added unit tests
Added integration tests

### CODE REVIEW

File: src/main/java/com/parkingapp/parkingservice/domain/vehicle/VehicleRepository.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package com.parkingapp.parkingservice.domain.vehicle;
+
+import java.util.Optional;
+import java.util.UUID;
+
+public interface VehicleRepository {
+    boolean saveVehicle(Vehicle vehicle);
+    Optional<Vehicle> getVehicleByUserIdAndPlate(UUID userId, String plate);
```
Comment (solblack) [inline src/main/java/com/parkingapp/parkingservice/domain/vehicle/VehicleRepository.java:8]:
```suggestion
    List<Vehicle> getUserVehicles(UUID userId);
```

File: src/main/java/com/parkingapp/parkingservice/domain/vehicle/Vehicle.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+@Data
+@AllArgsConstructor
+public class Vehicle {
+    private UUID id;
+    private String brand;
+    private String model;
+    private String color;
+    private String plate;
+    private String country;
```
Comment (solblack) [inline src/main/java/com/parkingapp/parkingservice/domain/vehicle/Vehicle.java:16]:
```suggestion
    private Country country;
```

and create enum Country with format: Country(String countryName, String isoCountryName, String iso2Name, int isoCode)

File: src/main/java/com/parkingapp/parkingservice/domain/vehicle/Vehicle.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+import java.util.UUID;
+
+@Data
+@AllArgsConstructor
+public class Vehicle {
+    private UUID id;
+    private String brand;
+    private String model;
+    private String color;
```
Comment (solblack) [inline src/main/java/com/parkingapp/parkingservice/domain/vehicle/Vehicle.java:14]: Maybe we could use an enum here, just the name e.g.: 

enum Color {
YELLOW, RED, PINK, BLACK
}

File: src/main/java/com/parkingapp/parkingservice/infrastructure/database/JdbcVehiclesRepository.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        String sql = """
+                SELECT * FROM vehicles
+                WHERE user_id = :userId AND plate = :plate
+                """;
+
+        SqlParameterSource params = new MapSqlParameterSource()
+                .addValue("userId", userId)
+                .addValue("plate", plate);
+
+        RowMapper<Vehicle> rowMapper = (rs, rowNum) -> new Vehicle(
```
Comment (solblack) [inline src/main/java/com/parkingapp/parkingservice/infrastructure/database/JdbcVehiclesRepository.java:52]: Extract mapper outside of method to be able to re use it by another method in the future

File: src/main/java/com/parkingapp/parkingservice/infrastructure/entrypoint/rest/VehiclesController.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            @ApiResponse(
+                    responseCode = "201",
+                    description = "Successful response",
+                    content = {
+                            @Content(
+                                    mediaType = "application/json",
+                                    schema = @Schema(implementation = ParkingResponse.class)
+                            )
+                    }),
+            @ApiResponse(
```
Comment (solblack) [inline src/main/java/com/parkingapp/parkingservice/infrastructure/entrypoint/rest/VehiclesController.java:46]: all ApiResponse doc for 409 code

File: src/main/java/com/parkingapp/parkingservice/infrastructure/entrypoint/rest/VehiclesController.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+                                    schema = @Schema(implementation = ErrorResponse.class)
+                            )
+                    }
+            )
+    })
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    public VehicleResponse createVehicle(@RequestBody @Valid CreateVehicleRequest request) {
+        Vehicle vehicle = new Vehicle(
+                UUID.randomUUID(),
```
Comment (solblack) [inline src/main/java/com/parkingapp/parkingservice/infrastructure/entrypoint/rest/VehiclesController.java:71]: Use IdGenerator

File: src/test/java/com/parkingapp/parkingservice/infrastructure/database/JdbcVehiclesRepositoryIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+                plate,
+                country,
+                userId
+        );
+
+        // WHEN
+        boolean saveResult = vehicleRepository.saveVehicle(newVehicle);
+        boolean saveResult2 = vehicleRepository.saveVehicle(duplicateVehicle);
+
+        // THEN
```
Comment (solblack) [inline src/test/java/com/parkingapp/parkingservice/infrastructure/database/JdbcVehiclesRepositoryIntegrationTest.java:151]: fetch user vehicles and check there's only 1 saved and it's the first one saved

File: src/test/java/com/parkingapp/parkingservice/infrastructure/database/JdbcVehiclesRepositoryIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+        Optional<Vehicle> result2 = vehicleRepository.getVehicleByUserIdAndPlate(
+                newVehicleUserIdAlreadyExists.getUserId(),
+                newVehicleUserIdAlreadyExists.getPlate()
+        );
+        assertThat(result2).isEqualTo(Optional.of(newVehicleUserIdAlreadyExists));
+        assertThat(saveResult2).isTrue();
+    }
+
+    @Test
+    void shouldNotSaveAVehicleWhenPlateAndUserIdExists() {
```
Comment (solblack) [inline src/test/java/com/parkingapp/parkingservice/infrastructure/database/JdbcVehiclesRepositoryIntegrationTest.java:135]:
```suggestion
    void shouldNotSaveAVehicleWhenItAlreadyExists() {
```

File: src/test/java/com/parkingapp/parkingservice/infrastructure/database/JdbcVehiclesRepositoryIntegrationTest.java (hunk 7 above)
Comment (solblack) [reply src/test/java/com/parkingapp/parkingservice/infrastructure/database/JdbcVehiclesRepositoryIntegrationTest.java:151]: this is to check that only 1 vehicle is saved per user id and plate and that the vehicle with same user id and plate is not wrongly updated

File: src/test/java/com/parkingapp/parkingservice/infrastructure/database/JdbcVehiclesRepositoryIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+        boolean saveResult = vehicleRepository.saveVehicle(newVehicle);
+
+        // THEN
+        Optional<Vehicle> result = vehicleRepository.getVehicleByUserIdAndPlate(newVehicle.getUserId(), newVehicle.getPlate());
+        assertThat(result).isEqualTo(Optional.of(newVehicle));
+        assertThat(saveResult).isTrue();
+    }
+
+    @Test
+    void shouldSaveVehicleWhenSamePlateAndDifferentUserIdExists() {
```
Comment (solblack) [inline src/test/java/com/parkingapp/parkingservice/infrastructure/database/JdbcVehiclesRepositoryIntegrationTest.java:69]:
```suggestion
    void shouldSaveVehicleWhenVehicleWithSamePlateAndDifferentUserIdExists() {
```

File: src/test/java/com/parkingapp/parkingservice/infrastructure/database/JdbcVehiclesRepositoryIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+        Optional<Vehicle> result2 = vehicleRepository.getVehicleByUserIdAndPlate(
+                newVehiclePlateAlreadyExists.getUserId(),
+                newVehiclePlateAlreadyExists.getPlate()
+        );
+        assertThat(result2).isEqualTo(Optional.of(newVehiclePlateAlreadyExists));
+        assertThat(saveResult2).isTrue();
+    }
+
+    @Test
+    void shouldSaveVehicleWhenSameUserIdAndDifferentPlateExists() {
```
Comment (solblack) [inline src/test/java/com/parkingapp/parkingservice/infrastructure/database/JdbcVehiclesRepositoryIntegrationTest.java:102]:
```suggestion
    void shouldSaveVehicleWhenVehicleWithSameUserIdAndDifferentPlateExists() {
```

File: src/test/java/com/parkingapp/parkingservice/infrastructure/database/JdbcVehiclesRepositoryIntegrationTest.java (hunk 9 above)
Comment (fabriciodigennaro) [reply src/test/java/com/parkingapp/parkingservice/infrastructure/database/JdbcVehiclesRepositoryIntegrationTest.java:69]: 👍🏻

File: src/test/java/com/parkingapp/parkingservice/infrastructure/database/JdbcVehiclesRepositoryIntegrationTest.java (hunk 10 above)
Comment (fabriciodigennaro) [reply src/test/java/com/parkingapp/parkingservice/infrastructure/database/JdbcVehiclesRepositoryIntegrationTest.java:102]: 👍🏻 +1

File: src/test/java/com/parkingapp/parkingservice/infrastructure/database/JdbcVehiclesRepositoryIntegrationTest.java (hunk 8 above)
Comment (fabriciodigennaro) [reply src/test/java/com/parkingapp/parkingservice/infrastructure/database/JdbcVehiclesRepositoryIntegrationTest.java:135]: 👍🏻

File: src/main/java/com/parkingapp/parkingservice/infrastructure/entrypoint/rest/GlobalExceptionHandler.java [hunk 11]
```diff
@@ -54,4 +55,10 @@ ErrorResponse handle(HttpMessageNotReadableException exception) {
         return new ErrorResponse(exception.getMessage());
     }
 
+    @ExceptionHandler(VehicleAlreadyExistsException.class)
+    @ResponseStatus(HttpStatus.CONFLICT)
```
[the remark this answers is not in the document]
Comment (fabriciodigennaro) [reply src/main/java/com/parkingapp/parkingservice/infrastructure/entrypoint/rest/GlobalExceptionHandler.java:59]: 😄

File: src/main/java/com/parkingapp/parkingservice/infrastructure/entrypoint/rest/VehiclesController.java [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+
+    @Operation(summary = "Create a new vehicle")
+    @ApiResponses(value = {
+            @ApiResponse(
+                    responseCode = "201",
+                    description = "Successful response",
+                    content = {
+                            @Content(
+                                    mediaType = "application/json",
+                                    schema = @Schema(implementation = ParkingResponse.class)
```
[the remark this answers is not in the document]
Comment (fabriciodigennaro) [reply src/main/java/com/parkingapp/parkingservice/infrastructure/entrypoint/rest/VehiclesController.java:43]: 🦝

File: src/main/java/com/parkingapp/parkingservice/infrastructure/entrypoint/rest/VehiclesController.java (hunk 5 above)
Comment (fabriciodigennaro) [reply src/main/java/com/parkingapp/parkingservice/infrastructure/entrypoint/rest/VehiclesController.java:46]: done

File: src/main/java/com/parkingapp/parkingservice/infrastructure/entrypoint/rest/VehiclesController.java (hunk 6 above)
Comment (fabriciodigennaro) [reply src/main/java/com/parkingapp/parkingservice/infrastructure/entrypoint/rest/VehiclesController.java:71]: Changed!

File: src/test/java/com/parkingapp/parkingservice/infrastructure/database/JdbcVehiclesRepositoryIntegrationTest.java (hunk 9 above)
Comment (fabriciodigennaro) [reply src/test/java/com/parkingapp/parkingservice/infrastructure/database/JdbcVehiclesRepositoryIntegrationTest.java:69]: done

File: src/main/java/com/parkingapp/parkingservice/domain/vehicle/Vehicle.java (hunk 2 above)
Comment (fabriciodigennaro) [reply src/main/java/com/parkingapp/parkingservice/domain/vehicle/Vehicle.java:16]: Added country enum!

File: src/main/java/com/parkingapp/parkingservice/infrastructure/database/JdbcVehiclesRepository.java (hunk 4 above)
Comment (fabriciodigennaro) [reply src/main/java/com/parkingapp/parkingservice/infrastructure/database/JdbcVehiclesRepository.java:52]: 👍🏻

File: src/main/java/com/parkingapp/parkingservice/domain/vehicle/VehicleRepository.java (hunk 1 above)
Comment (fabriciodigennaro) [reply src/main/java/com/parkingapp/parkingservice/domain/vehicle/VehicleRepository.java:8]: Changed to new method in repository!

File: src/test/java/com/parkingapp/parkingservice/infrastructure/database/JdbcVehiclesRepositoryIntegrationTest.java (hunk 7 above)
Comment (fabriciodigennaro) [reply src/test/java/com/parkingapp/parkingservice/infrastructure/database/JdbcVehiclesRepositoryIntegrationTest.java:151]: done! 💯

File: src/main/java/com/parkingapp/parkingservice/domain/exceptions/VehicleAlreadyExistsException.java [hunk 13]
```diff
@@ -0,0 +1,7 @@
+package com.parkingapp.parkingservice.domain.exceptions;
+
+public class VehicleAlreadyExistsException extends RuntimeException {
+    public VehicleAlreadyExistsException(String message) {
```
Comment (solblack) [inline src/main/java/com/parkingapp/parkingservice/domain/exceptions/VehicleAlreadyExistsException.java:4]:
```suggestion
    public VehicleAlreadyExistsException() {
```

File: src/main/java/com/parkingapp/parkingservice/domain/exceptions/VehicleAlreadyExistsException.java [hunk 14]
```diff
@@ -0,0 +1,7 @@
...
+        super(message);
```
Comment (solblack) [inline src/main/java/com/parkingapp/parkingservice/domain/exceptions/VehicleAlreadyExistsException.java:5]:
```suggestion
        super("The combination of vehicle_id and user_id already exists.");
```

File: src/main/java/com/parkingapp/parkingservice/domain/vehicle/Vehicle.java [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+
+import java.util.UUID;
+
+@Data
+@AllArgsConstructor
+public class Vehicle {
+    private UUID id;
+    private String brand;
+    private String model;
+    private VehicleColor color;
```
Comment (solblack) [inline src/main/java/com/parkingapp/parkingservice/domain/vehicle/Vehicle.java:15]:
```suggestion
    private Color color;
```

**Review by solblack — APPROVED** [review]

**Pull request merged**